### PR TITLE
feat(flutter_test): Adding Roboto as font fol golden tests

### LINKED ISSUE #95733: feat(flutter_test): Add Roboto as default other than Ahem for golden tests
## Use case

Golden tests are great but they use a default font that replaces actual characters with squares (Ahem font).

In order to make goldens more valuable, it would be great if there was the possibility to use another more meaningful font (like Roboto). It would make the text actually rendere characters instead of squares.

In our projects we use goldens a lot but we need to rely on workarounds or third-party libraries to make characters correctly display

## Proposal

We currently generate goldens with

```dart
await expectLater(
  find.byType(SomeWidgetWithText),
  matchesGoldenFile('some_widget_with_test'),
);
```

I would like to open a PR to add the possibility to use another param:

```dart
await expectLater(
  find.byType(SomeWidgetWithText),
  matchesGoldenFile('some_widget_with_test', useRoboto: true),
);
```

The `useRoboto` would be `false` by default to preserve the current behavior but, when `true`, it would use `Roboto` (or any other available font) instead of Ahem. May this work?

### PR DESCRIPTION
**Pull request by albertodev01**
This PR addresses the #95733 issue. Golden tests always use the Ahem font but sometimes we may want to make images more valuable by using a human-readable font. For this reason, I have added an optional `useRoboto` font which loads Roboto instead of Ahem for golden tests.

 - Fixes #95733 

I have made a change in the `flutter_test` package but it's not a breaking one. The default behavior is preserved since

```dart
await expectLater(
  someFinder,
  matchesGoldenFile('image.png'),
);
```

still works as before. It's only been added a new parameter (which is `false` by default) that may replace Ahem with Roboto:

```dart
await expectLater(
  someFinder,
  matchesGoldenFile('image.png', useRoboto: true),
);
```

With this new config, icons and text will render using Roboto.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

### CODE REVIEW

**Comment by albertodev01** [comment]
@Piinks tagging you here for a review as agreed on Discord!

**Comment by goderbauer** [comment]
It seems strange to hard-code Roboto. Why that particular font? What if my app uses different fonts?

**Comment by albertodev01** [comment]
> It seems strange to hard-code Roboto. Why that particular font? What if my app uses different fonts?

I thiught about it because it is Flutter's default font for Android. There are a few alternatives I think:

  1. Name the param `usePlatformFont` and then internally I would decide which font to load (like `Roboto` for Android, `SanFrancisco` for iOS and so on...)
  2. Do the above plus adding a `useCustomFont` callback so that you can load your own font
  3. Only have a `useCustomFont` param for users to load a custom font on their own. In addition, I would add some "default" public callbacks like `useRoboto` or `useSanFrancisco` so users can quickly use them without having to write a function on their own

I think that (3) would be the most flessibile one. Maybe I could make an `abstract class CustomFontLoader` which then provider some pre-defined ones for Roboto and SanFrancisco but then anyone can create whatever they want.

What do you think? @goderbauer

**Comment by goderbauer** [comment]
I don't think that hard-coding this into the golden functionality is the right way. If you want to take golden images with real fonts, you can already do that by just loading the font(s) you need in the test (see also https://stackoverflow.com/questions/62447898/flutter-widget-test-cannot-emulate-different-screen-size-properly/62460566#62460566). 

What we are perhaps missing is some documentation on the golden matchers that explain how to do this. We could potentially provide a wrapping function that encapsulates more of the font loading, but it shouldn't assume any default fonts.

File: packages/flutter_test/test/matchers_test.dart [hunk 1]
```diff
@@ -363,6 +363,15 @@ void main() {
         expect(comparator.golden, Uri.parse('foo.png'));
       });
 
+      testWidgets('if comparator succeeds using Roboto', (WidgetTester tester) async {
+        await tester.pumpWidget(boilerplate(const Text('hello')));
+        final Finder finder = find.byType(Text);
+        await expectLater(finder, matchesGoldenFile('foo.png', useRoboto: true,));
```
Comment (Piinks) [inline packages/flutter_test/test/matchers_test.dart:369]: I don't think we have golden file tests in the flutter_test package currently. I am surprised this isn't failing.

**Review by Piinks — COMMENTED** [review]
Have you tried this across different platforms to see how much the images vary with the font? I think the reason why we sub in the ahem font is primarily because very small rendering differences across platforms cause the golden file image tests to be very brittle, almost useless if development is occurring across different machines.

There is a package that allows you to load fonts, and in that case you could use any font you want: https://pub.dev/packages/golden_toolkit

I agree with @goderbauer, I don't think this is the right approach.

**Comment by albertodev01** [comment]
I think I'm going this way:

> What we are perhaps missing is some documentation on the golden matchers that explain how to do this.

For sure an example in the docstrings would help people. Thank you for your time in this review @goderbauer 

---

@Piinks thanks you too for your effort in the review. I get your point, I think I should go for another strategy. Following this advice

> We could potentially provide a wrapping function that encapsulates more of the font loading

I think that I could do something like this:

```dart
await expectLater(
  someFinder,
  matchesGoldenFile(
    'image.png', 
    customFont: () {
      // This can be ANY font loading logic for the golden
      return rootBundle.load('assets/fonts/my_font.ttf');
    },
  ),
);
```

To make fonts loading easier, I could create a `customFont` callback where the developer can load **any** font (setup and `FontLoader` calls will be handled internally). 

May this work better? Before I take this way, would you please recommend me if it sounds a good idea?

**Comment by Piinks** [comment]
I think that sounds good, thanks for working on this! :)

**Comment by goderbauer** [comment]
Please do not add this to matchesGoldenFile. What I meant is the testing library could maybe provide a convenience function to load fonts in general. This shouldn't be baked into the golden logic. But as @Piinks it's not really needed because https://pub.dev/packages/golden_toolkit#loading-fonts already provides something like that. I think we are just missing docs here.

**Comment by albertodev01** [comment]
Following your suggestion, I have closed this PR and opened another [one](https://github.com/flutter/flutter/pull/96194) where I just improve the documentation

**Pull request closed without merging**